cas-server-core/src/main/java/org/jasig/cas/authentication/support/DefaultCasAttributeEncoder.java

```diff
@@ -0,0 +1,227 @@
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);
+    }
+
+    /**
+     * Encrypt, encode and put the attribute into attributes map.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param cachedAttributeName the cached attribute name
+     */
+    protected final void encryptAndEncodeAndPutIntoAttributesMap(final Map<String, Object> attributes,
+                                                           final Map<String, String> cachedAttributesToEncode,
+                                                           final String cachedAttributeName) {
+        final String cachedAttribute = cachedAttributesToEncode.remove(cachedAttributeName);
+        if (StringUtils.isNotBlank(cachedAttribute)) {
+            logger.debug("Retrieved [{}] as a cached model attribute...", cachedAttributeName);
+            final String encodedValue = CompressionUtils.encryptAndEncodeBase64(cachedAttribute, this.cipher);
+            attributes.put(cachedAttributeName, encodedValue);
+            logger.debug("Encrypted and encoded [{}] as an attribute to [{}].", cachedAttributeName, encodedValue);
+        } else {
+            logger.debug("[{}] is not available as a cached model attribute to encrypt...", cachedAttributeName);
+        }
+    }
+
+    /**
+     * Initialize the cipher with the public key
+     * and then start to encrypt select attributes.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param service the service
+     */
+    protected void encodeAttributesInternal(final Map<String, Object> attributes,
+                                            final Map<String, String> cachedAttributesToEncode,
+                                            final RegisteredService service) {
+        encodeAndEncryptCredentialPassword(attributes, cachedAttributesToEncode);
+        encodeAndEncryptProxyGrantingTicket(attributes, cachedAttributesToEncode);
+    }
+
+    /**
+     * Initialize cipher based on service public key.
+     *
+     * @param service the service
+     * @return the false if no public key is found
+     * or if cipher cannot be initialized, etc.
+     */
+    protected final boolean initializeCipherBasedOnServicePublicKey(final RegisteredService service) {
+        try {
+            final PublicKey publicKey = service.getPublicKey().createInstance();
+            if (publicKey == null) {
+                logger.debug("No public key is defined for service [{}]. No encoding will take place.", service);
+                return false;
+            }
+            logger.debug("Using public key [{}] to initialize the cipher", service.getPublicKey());
+
+            this.cipher.init(Cipher.ENCRYPT_MODE, publicKey);
```

Comment:
This appears to be a thread safety problem. This component appears to be wired into the Spring context as a singleton (protocolViewsConfiguration.xml), so this code could be called by multiple threads afaict. If that analysis is correct, then this code will break under concurrency; Cipher instances are not thread safe. I would recommend initializing a local Cipher instance here on demand. Alternatively, I'm open to a feature request to create an implementation of https://github.com/vt-middleware/cryptacular/blob/master/src/main/java/org/cryptacular/bean/CipherBean.java that uses asymmetric keys. That component is much more Spring friendly in terms of cipher parameter specification and is thread safe.